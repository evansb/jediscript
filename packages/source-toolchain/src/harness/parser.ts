import { createContext } from '../context'
import { parse, FatalSyntaxError } from '../parser'

type Options = {
  week: number
  errorClass: Function
  explanation: RegExp
}

const defaultOptions: Options = {
  week: 3,
  errorClass: FatalSyntaxError,
  explanation: /.*/
}

export const singleError = (source: string, options: Partial<Options>) => {
  const completeOptions: Options = {
    ...defaultOptions,
    ...options
  }
  const { errorClass, week, explanation } = completeOptions
  const context = createContext({ week })
  parse(source, context)
  expect(context.parser.errors.length).toBe(1)
  expect(context.parser.errors[0]).toBeInstanceOf(errorClass)
  expect(context.parser.errors[0].explain()).toMatch(explanation)
}
