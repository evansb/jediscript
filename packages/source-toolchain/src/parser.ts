/// <reference path='parser.d.ts' />
import * as es from 'estree'
import {
  parse as acornParse,
  Options as AcornOptions,
  SourceLocation,
  Position,
} from 'acorn'
import { simple } from 'acorn/dist/walk'
import { ErrorType } from './types/error'
import { StaticState } from './types/static'
import syntaxTypes from './syntaxTypes'
import { createContext } from './context'

export type ParserOptions = {
  week: number,
}

export const freshId = (() => {
  let id = 0
  return () => {
    id++
    return 'node_' + id
  }
})()

function compose<T extends es.Node, S>(
  w1: (node: T, state: S) => void,
  w2: (node: T, state: S) => void) {
  return (node: T, state: S) => {
    w1(node, state)
    w2(node, state)
  }
}

const walkers: {[name: string]: (node: es.Node, state: StaticState) => void } = {}

for (const type of Object.keys(syntaxTypes)) {
  walkers[type] = (node: es.Node, state: StaticState) => {
    const id = freshId()
    Object.defineProperty(node, '__id', {
      enumerable: true,
      configurable: false,
      writable: false,
      value: id,
    })
    state.cfg.nodes[id] = {
      node,
      scope: undefined,
      edges: [],
      usages: [],
    }
    if (syntaxTypes[node.type] > state.week) {
      state.parser.errors.push({
        kind: 'syntax',
        type: ErrorType.MatchFailure,
        node,
      })
    }
  }
}

// If Statement must
// 1. Have Else case (week <= 3)
// 2. If and Els case must be Surrounded by braces
const checkIfStatement = (node: es.IfStatement, state: StaticState) => {
  if (node.consequent! && node.consequent.type !== 'BlockStatement') {
    state.parser.errors.push({
      kind: 'syntax',
      type: ErrorType.IfConsequentNotABlockStatement,
      node,
    })
  }
  if (state.week <= 3 && !node.alternate) {
    state.parser.errors.push({
      kind: 'syntax',
      type: ErrorType.MissingIfAlternate,
      node,
    })
  } else if (node.alternate && node.alternate.type !== 'BlockStatement') {
    state.parser.errors.push({
      kind: 'syntax',
      type: ErrorType.IfAlternateNotABlockStatement,
      node,
    })
  }
}
walkers.IfStatement = compose(walkers.IfStatement, checkIfStatement)

// Binary Expressions
// == and != are banned, must use !== and ===
const checkBinaryExpression = (node: es.BinaryExpression, state: StaticState) => {
  if (node.operator === '==') {
    state.parser.errors.push({
      kind: 'syntax',
      type: ErrorType.UseStrictEquality,
      node,
    })
  } else if (node.operator === '!=') {
    state.parser.errors.push({
      kind: 'syntax',
      type: ErrorType.UseStrictInequality,
      node,
    })
  }
}
walkers.BinaryExpression = compose(walkers.BinaryExpression, checkBinaryExpression)

// Variable Declarations
// Can only have single declarations
const checkVariableDeclaration = (node: es.VariableDeclaration, state: StaticState) => {
  if (node.declarations.length > 1) {
    state.parser.errors.push({
      kind: 'syntax',
      type: ErrorType.MultipleDeclarations,
      node,
    })
  } else if (node.declarations.length == 1 && !node.declarations[0].init) {
    state.parser.errors.push({
      kind: 'syntax',
      type: ErrorType.NoDeclarations,
      node,
    })
  }
}
walkers.VariableDeclaration = compose(walkers.VariableDeclaration, checkVariableDeclaration)

const createAcornParserOptions = (state: StaticState): AcornOptions => ({
  sourceType: 'script',
  ecmaVersion: 5,
  locations: true,
  onInsertedSemicolon(end: any, loc: any) {
    const node = ({
      type: 'Statement',
      loc: {
        end: {line: loc.line, column: loc.column + 1},
        start: loc,
      },
    }) as any
    state.parser.errors.push({
      kind: 'syntax',
      type: ErrorType.MissingSemicolon,
      node,
    })
  },
  onTrailingComma(end: any, loc: Position) {
    const node = ({
      type: 'Statement',
      loc: {
        end: { line: loc.line, column: loc.column + 1 },
        start: loc,
      },
    }) as any
    state.parser.errors.push({
      kind: 'syntax',
      type: ErrorType.TrailingComma,
      node,
    })
  },
  onComment: state.parser.comments,
})

export const parse = (source: string, state: StaticState | number) => {
  if (typeof state === 'number') {
    state = createContext({ week: state })
  }
  try {
    const program = acornParse(source, createAcornParserOptions(state))
    state.parser.program = program
    simple(program, walkers, undefined, state)
  } catch (error) {
    if (error instanceof SyntaxError) {
      const loc = (error as any).loc
      state.parser.errors.push({
        kind: 'syntax',
        type: ErrorType.AcornParseError,
        explanation: error.toString(),
        node: ({
          type: 'Statement',
          loc: {
            start: { line: loc.line, column: loc.column },
            end: { line: loc.line, column: loc.column + 1 },
          },
        }) as any,
      })
    } else {
      throw error
    }
  }
  return state
}
