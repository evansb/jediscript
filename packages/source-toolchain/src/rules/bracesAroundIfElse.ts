import * as es from 'estree'
import { IError } from '../types/error'
import { Rule } from '../types/static'

export class BracesAroundIfElseError implements IError {
  constructor(
    public node: es.IfStatement,
    public type: 'consequent' | 'alternate'
  ) {}

  get location() {
    return this.node.loc!
  }

  explain() {
    if (this.type === 'consequent') {
      return 'Missing curly braces around "if"'
    } else {
      return 'Missing curly braces around "else"'
    }
  }

  elaborate() {
    return 'TODO'
  }
}

const bracesAroundIfElse: Rule<es.IfStatement> = {
  name: 'braces-around-if-else',

  checkNodes: {
    IfStatement(node: es.IfStatement) {
      const errors: IError[] = []
      if (node.consequent && node.consequent.type !== 'BlockStatement') {
        errors.push(new BracesAroundIfElseError(node, 'consequent'))
      }
      if (node.alternate && node.alternate.type !== 'BlockStatement') {
        errors.push(new BracesAroundIfElseError(node, 'alternate'))
      }
      return errors
    }
  }
}

export default bracesAroundIfElse
