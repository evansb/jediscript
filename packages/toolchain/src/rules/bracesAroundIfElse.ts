import * as es from 'estree'
import { generate } from 'astring'
import { stripIndent } from 'common-tags'
import { IError } from '../types/error'
import { Rule } from '../types/static'

export class BracesAroundIfElseError implements IError {
  constructor(
    public node: es.IfStatement,
    public type: 'consequent' | 'alternate'
  ) {}

  get location() {
    return this.node.loc!
  }

  explain() {
    if (this.type === 'consequent') {
      return 'Missing curly braces around "if" block'
    } else {
      return 'Missing curly braces around "else" block'
    }
  }

  elaborate() {
    let ifOrElse
    let header
    let body
    if (this.type === 'consequent') {
      ifOrElse = 'if'
      header = `if (${generate(this.node.test)})`
      body = this.node.consequent
    } else {
      ifOrElse = header = 'else'
      body = this.node.alternate
    }

    return stripIndent`
      ${ifOrElse} block need to be enclosed with a pair of curly braces.

      ${header} {
        ${generate(body)}
      }

      An exception is when you have an "if" followed by "else if", in this case
      "else if" block does not need to be surrounded by curly braces.

      if (someCondition) {
        // ...
      } else /* notice missing { here */ if (someCondition) {
        // ...
      } else {
        // ...
      }

      Rationale: Readability in dense packed code.

      In the snippet below, for instance, with poor indentation it is easy to
      mistaken hello() and world() to belong to the same branch of logic.

      if (someCondition) {
        2;
      } else
        hello();
      world();

    `
  }
}

const bracesAroundIfElse: Rule<es.IfStatement> = {
  name: 'braces-around-if-else',

  checkNodes: {
    IfStatement(node: es.IfStatement) {
      const errors: IError[] = []
      if (node.consequent && node.consequent.type !== 'BlockStatement') {
        errors.push(new BracesAroundIfElseError(node, 'consequent'))
      }
      if (node.alternate) {
        const notBlock = node.alternate.type !== 'BlockStatement'
        const notIf = node.alternate.type !== 'IfStatement'
        if (notBlock && notIf) {
          errors.push(new BracesAroundIfElseError(node, 'alternate'))
        }
      }
      return errors
    }
  }
}

export default bracesAroundIfElse
