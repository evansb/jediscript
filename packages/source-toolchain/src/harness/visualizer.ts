import { generate } from 'escodegen'
import { parse } from '../parser'
import { createState, evalProgram } from '../evaluator'
import { create, next } from '../visualizer'

export const testVisualizer = (program: string, expectedOutput: string[]) => {
  let state = createState()
  let visualizer = create()

  const node = parse(program, 3).node!
  const evaluator = evalProgram(node, state)
  const prints = []

  let gen = evaluator.next()
  while (gen) {
    state = gen.value
    if (!state) { break }
    const prev = visualizer
    visualizer = next(visualizer, state)
    if (visualizer.root && (visualizer.id !== prev.id)) {
      prints.push(generate(visualizer.root))
    }
    gen = evaluator.next()
  }
  expect(prints).toEqual(expectedOutput)
}
