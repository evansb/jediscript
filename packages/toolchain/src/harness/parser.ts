import { createContext } from '../context'
import { parse, FatalSyntaxError } from '../parser'

type Options = {
  week: number
  errorClass: any
}

const defaultOptions: Options = {
  week: 3,
  errorClass: FatalSyntaxError
}

const runParser = (source: string, week: number) => {
  const context = createContext({ week })
  parse(source, context)
  return context
}

export const singleError = (source: string, options: Partial<Options>) => {
  const completeOptions: Options = {
    ...defaultOptions,
    ...options
  }
  const { errorClass, week } = completeOptions
  const context = runParser(source, week)
  expect(context.parser.errors.length).toBe(1)
  expect(context.parser.errors[0]).toBeInstanceOf(errorClass)
  expect(context.parser.errors[0].explain()).toMatchSnapshot('explanation')
  expect(context.parser.errors[0].elaborate()).toMatchSnapshot('elaboration')
}

export const noError = (source: string, options: Partial<Options> = {}) => {
  const completeOptions: Options = {
    ...defaultOptions,
    ...options
  }
  const { week } = completeOptions
  const context = runParser(source, week)
  expect(context.parser.errors.length).toBe(0)
}
