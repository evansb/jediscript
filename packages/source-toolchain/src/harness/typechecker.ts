import { createContext } from '../../src/context'
import { CFG, StaticState } from '../../src/types/static'
import { isSameType, parseString } from '../../src/typechecker'

type Suite = {
  name: string
  source: string
  assertions: Array<{ name: string; type: CFG.Type }>
}

export const parseTypecheckerTest = (testFileContent: string): Suite[] => {
  const lines = testFileContent.split(/\n/)
  const suites: Suite[] = []

  let source = ''
  let testName = ''
  let assertions: Array<{ name: string; type: CFG.Type }> = []

  for (let line of lines) {
    line = line.trim()

    // Start of a new test
    // Collect previous suite
    if (line.match(/\/\/@/) && source.length > 0) {
      suites.push({
        name: testName,
        source,
        assertions
      })
      source = ''
      assertions = []
    }

    if (line.match(/\/\/!/)) {
      const assertion = line.split(/\/\/!/)[1]
      const [name, typeString] = assertion.trim().split(/:/)
      assertions.push({
        name: name.trim(),
        type: parseString(typeString.trim())
      })
    } else if (line.match(/\/\/@/)) {
      testName = line.split(/\/\/@/)[1].trim()
    }
    if (line) {
      source += line
      source += '\n'
    }
  }

  // Collect last suite
  if (source.length > 0) {
    suites.push({
      name: testName,
      source,
      assertions
    })
  }

  return suites
}

export const runTypecheckerTest = (
  suites: Suite[],
  parse: (source: string, state: StaticState) => void,
  generateCFG: (state: StaticState) => void,
  typecheck: (state: StaticState) => void
) => {
  suites.forEach(s => {
    const state = createContext({ week: 3 })
    parse(s.source, state)
    generateCFG(state)
    typecheck(state)
    s.assertions.forEach(a => {
      const { name, type: expectedType } = a
      expect(isSameType(state.cfg.scopes[0].env[name].type, expectedType)).toBe(
        true
      )
    })
  })
}
