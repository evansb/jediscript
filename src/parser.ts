/// <reference path='parser.d.ts' />
import * as es from 'estree'
import {
  parse as acornParse,
  Options as AcornOptions,
  SourceLocation,
  Position
} from 'acorn'
import { stripIndent } from 'common-tags'
import { simple } from 'acorn/dist/walk'

import { IError } from './types/error'
import { StaticState } from './types/static'
import syntaxTypes from './syntaxTypes'
import { createContext } from './context'
import rules from './rules'

export type ParserOptions = {
  week: number
}

export class DisallowedConstructError implements IError {
  nodeType: string

  constructor(public node: es.Node) {
    this.nodeType = this.splitNodeType()
  }

  get location() {
    return this.node.loc!
  }

  explain() {
    return `${this.nodeType} is not allowed`
  }

  elaborate() {
    return stripIndent`
      You are trying to use ${this.nodeType}, which is not yet allowed (yet).
    `
  }

  private splitNodeType() {
    const nodeType = this.node.type
    const tokens: string[] = []
    let soFar = ''
    for (let i = 0; i < nodeType.length; i++) {
      const isUppercase = nodeType[i] === nodeType[i].toUpperCase()
      if (isUppercase && i > 0) {
        tokens.push(soFar)
        soFar = ''
      } else {
        soFar += nodeType[i]
      }
    }
    return tokens.join(' ')
  }
}

export class FatalSyntaxError implements IError {
  constructor(public location: es.SourceLocation, public message: string) {}

  explain() {
    return `Syntax Error: ${this.message}`
  }

  elaborate() {
    return 'There is a syntax error in your program'
  }
}

export class MissingSemicolonError implements IError {
  constructor(public location: es.SourceLocation) {}

  explain() {
    return 'Missing semicolon at the end of statement'
  }

  elaborate() {
    return 'Every statement must be terminated by a semicolon.'
  }
}

export class TrailingCommaError implements IError {
  constructor(public location: es.SourceLocation) {}

  explain() {
    return 'Trailing comma'
  }

  elaborate() {
    return 'Please remove the trailing comma'
  }
}

export const freshId = (() => {
  let id = 0
  return () => {
    id++
    return 'node_' + id
  }
})()

function compose<T extends es.Node, S>(
  w1: (node: T, state: S) => void,
  w2: (node: T, state: S) => void
) {
  return (node: T, state: S) => {
    w1(node, state)
    w2(node, state)
  }
}

const walkers: {
  [name: string]: (node: es.Node, state: StaticState) => void
} = {}

for (const type of Object.keys(syntaxTypes)) {
  walkers[type] = (node: es.Node, state: StaticState) => {
    const id = freshId()
    Object.defineProperty(node, '__id', {
      enumerable: true,
      configurable: false,
      writable: false,
      value: id
    })
    state.cfg.nodes[id] = {
      id,
      node,
      scope: undefined,
      usages: []
    }
    state.cfg.edges[id] = []
    if (syntaxTypes[node.type] > state.week) {
      state.parser.errors.push(new DisallowedConstructError(node))
    }
  }
}

const createAcornParserOptions = (state: StaticState): AcornOptions => ({
  sourceType: 'script',
  ecmaVersion: 5,
  locations: true,
  onInsertedSemicolon(end: any, loc: any) {
    state.parser.errors.push(
      new MissingSemicolonError({
        end: { line: loc.line, column: loc.column + 1 },
        start: loc
      })
    )
  },
  onTrailingComma(end: any, loc: Position) {
    state.parser.errors.push(
      new TrailingCommaError({
        end: { line: loc.line, column: loc.column + 1 },
        start: loc
      })
    )
  },
  onComment: state.parser.comments
})

rules.forEach(rule => {
  const keys = Object.keys(rule.checkNodes)
  keys.forEach(key => {
    walkers[key] = compose(walkers[key], (node, state) => {
      const checker = rule.checkNodes[key]
      const errors = checker(node)
      if (errors.length > 0) {
        state.parser.errors = state.parser.errors.concat(errors)
      }
    })
  })
})

export const parse = (source: string, state: StaticState | number) => {
  if (typeof state === 'number') {
    state = createContext({ week: state })
  }
  try {
    const program = acornParse(source, createAcornParserOptions(state))
    state.parser.program = program
    state.cfg.scopes[0].node = program
    simple(program, walkers, undefined, state)
  } catch (error) {
    if (error instanceof SyntaxError) {
      const loc = (error as any).loc
      const location = {
        start: { line: loc.line, column: loc.column },
        end: { line: loc.line, column: loc.column + 1 }
      }
      const message = error.toString()
      state.parser.errors.push(new FatalSyntaxError(location, error.toString()))
    } else {
      throw error
    }
  }
  return state
}
