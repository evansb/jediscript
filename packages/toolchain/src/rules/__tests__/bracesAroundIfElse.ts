import { singleError, noError } from '../../harness/parser'
import { BracesAroundIfElseError } from '../bracesAroundIfElse'

it('detects If not using curly braces', () => {
  singleError(
    `
    if (2 === 2)
      var zomg = 2;
    else {
      1 + 2;
    }
  `,
    {
      errorClass: BracesAroundIfElseError
    }
  )
})

it('detects Else not using curly braces', () => {
  singleError(
    `
    if (2 === 2) {
      1 + 2;
    } else
      var zomg = 2;
  `,
    {
      errorClass: BracesAroundIfElseError
    }
  )
})

it('allows else-if not using curly braces', () => {
  noError(
    `
    if (2 === 2) {
      1 + 1;
    } else if (1 === 2) {
      1 + 2;
    } else {
      1 + 3;
    }
  `
  )
})
