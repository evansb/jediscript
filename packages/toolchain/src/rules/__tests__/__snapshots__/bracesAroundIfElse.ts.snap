// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`elaboration 1`] = `
"if block need to be enclosed with a pair of curly braces.

if (2 === 2) {
  var zomg = 2;
}

An exception is when you have an \\"if\\" followed by \\"else if\\", in this case
\\"else if\\" block does not need to be surrounded by curly braces.

if (someCondition) {
  // ...
} else /* notice missing { here */ if (someCondition) {
  // ...
} else {
  // ...
}

Rationale:
Readability in dense packed code.
In the snippet below, for instance, with poor indentation it is easy to
mistaken hello() and world() to belong to the same branch of logic.

if (someCondition) {
  2;
} else
  hello();
world();"
`;

exports[`elaboration 2`] = `
"else block need to be enclosed with a pair of curly braces.

else {
  var zomg = 2;
}

An exception is when you have an \\"if\\" followed by \\"else if\\", in this case
\\"else if\\" block does not need to be surrounded by curly braces.

if (someCondition) {
  // ...
} else /* notice missing { here */ if (someCondition) {
  // ...
} else {
  // ...
}

Rationale:
Readability in dense packed code.
In the snippet below, for instance, with poor indentation it is easy to
mistaken hello() and world() to belong to the same branch of logic.

if (someCondition) {
  2;
} else
  hello();
world();"
`;

exports[`explanation 1`] = `"Missing curly braces around \\"if\\" block"`;

exports[`explanation 2`] = `"Missing curly braces around \\"else\\" block"`;
